p2p/transport.go

```diff
@@ -0,0 +1,579 @@
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p/conn"
+)
+
+const (
+	defaultDialTimeout      = time.Second
+	defaultFilterTimeout    = 5 * time.Second
+	defaultHandshakeTimeout = 3 * time.Second
+)
+
+// Transport-level errors.
+var (
+	ErrTransportFilterTimeout = errors.New("peer filter timeout")
+)
+
+// accept is the container to carry the upgraded connection and NodeInfo from an
+// asynchronously running routine to the Accept method.
+type accept struct {
+	conn     net.Conn
+	nodeInfo NodeInfo
+	err      error
+}
+
+// peerConfig is used to bundle data we need to fully setup a Peer with an
+// MConn, provided by the caller of Accept and Dial.
+type peerConfig struct {
+	chDescs              []*conn.ChannelDescriptor
+	onPeerError          func(Peer, interface{})
+	outbound, persistent bool
+	reactorsByCh         map[byte]Reactor
+}
+
+// Transport emits and connects to Peers. The implementation of Peer is left to
+// the transport. Each transport is also responsible to filter establishing
+// peers specific to its domain.
+type Transport interface {
+	// Accept returns a newly connected Peer.
+	Accept(peerConfig) (Peer, error)
+
+	// Dial connects to the Peer for the address.
+	Dial(NetAddress, peerConfig) (Peer, error)
+}
+
+// transportLifecycle bundles the methods for callers to control start and stop
+// behaviour.
+type transportLifecycle interface {
+	Close() error
+	Listen(NetAddress) error
+}
+
+// ConnFilterFunc to be implemented by filter hooks after a new connection has
+// been established.
+type ConnFilterFunc func(map[string]net.Conn, net.Conn) error
+
+// PeerFilterFunc to be implemented by filter hooks after a new Peer has been
+// fully setup.
+type PeerFilterFunc func(map[ID]Peer, Peer) error
+
+// IPResolver is a behaviour subset of net.Resolver.
+type IPResolver interface {
+	LookupIPAddr(context.Context, string) ([]net.IPAddr, error)
+}
+
+// ConnDuplicateIPFilter resolves and keeps all ips for an incoming connection
+// and refuses new ones if they come from a known ip.
+func ConnDuplicateIPFilter(resolver IPResolver) ConnFilterFunc {
+	im := map[string][]net.IP{}
+
+	return func(cs map[string]net.Conn, c net.Conn) error {
+		// Remove cached ips for dropped conns.
+		for addr := range im {
+			if _, ok := cs[addr]; !ok {
+				delete(im, addr)
+			}
+		}
+
+		// Resolve ips for incoming conn.
+		host, _, err := net.SplitHostPort(c.RemoteAddr().String())
+		if err != nil {
+			return err
+		}
+
+		addrs, err := resolver.LookupIPAddr(context.Background(), host)
+		if err != nil {
+			return err
+		}
+
+		newIPs := []net.IP{}
+
+		for _, addr := range addrs {
+			newIPs = append(newIPs, addr.IP)
+		}
+
+		for _, knownIPs := range im {
+			for _, known := range knownIPs {
+				for _, new := range newIPs {
+					if new.Equal(known) {
+						return &ErrRejected{
+							conn:        c,
+							err:         fmt.Errorf("IP<%v> already connected", new),
+							isDuplicate: true,
```

Comment:
doesn't really seem right for filters to be expected to return ErrRejected if we want to support filters implemented outside this package, they won't have access to `is*` variables

Reply:
It doesn't seem to interfer if filters provided by the `p2p` package return specialised errors. Alternatively we can think of exposing a way for callers to create the right errors.